## Exercise 3.1
In this exercise your task is to open and explore a NOAA weather data file using Pandas. The data file name is 6153237444115dat.csv and it is located in the data folder (add link). An overview of the tasks in this exercise:

Import the Pandas module
Read the data using Pandas into a variable called data
Calculate a number of basic statistics from the data


## Problem 1 - Read the file and clean it
Import the pandas module and read the weather data into a variable called data. Print the first five rows of the data file.

In [1]:
import pandas as pd
import numpy as np # this line was added

In [2]:
data = pd.read_csv('https://raw.githubusercontent.com/PhamNguyen97/Data_analysis_Rovaniemi_-_Helsinki-Kumpula_temperature/master/6153237444115dat.csv')

In [3]:
data.head()

USAF   WBAN  YR--MODAHRMN  DIR SPD  GUS  CLG  SKC  L  M  ...     SLP  \
0  28450  99999  201705010000  174  10   14  ***  ***  *  *  ...  1009.2   
1  28450  99999  201705010020  180  10  ***    4  ***  *  *  ...  ******   
2  28450  99999  201705010050  190  10  ***    4  ***  *  *  ...  ******   
3  28450  99999  201705010100  188  12   16  ***  ***  *  *  ...  1009.1   
4  28450  99999  201705010120  200  13  ***    2  OBS  *  *  ...  ******   

     ALT     STP  MAX  MIN  PCP01  PCP06  PCP24  PCPXX  SD  
0  *****   984.1  ***  ***  *****  *****  *****  *****  35  
1  29.74  ******  ***  ***  *****  *****  *****  *****  **  
2  29.74  ******  ***  ***  *****  *****  *****  *****  **  
3  *****   984.0  ***  ***  *****  *****  *****  *****  35  
4  29.74  ******  ***  ***  *****  *****  *****  *****  **  

[5 rows x 33 columns]

## Problem 2 - Basic characteristics of the data
Based on the data DataFrame from Problem 1, answer to following questions:

How many rows is there in the data?

What are the column names?

What are the datatypes of the columns?

In [4]:
n_rows = print(len(data))

11694


In [5]:
column_names = print(data.columns)

Index(['USAF', 'WBAN', 'YR--MODAHRMN', 'DIR', 'SPD', 'GUS', 'CLG', 'SKC', 'L',
       'M', 'H', 'VSB', 'MW', 'MW.1', 'MW.2', 'MW.3', 'AW', 'AW.1', 'AW.2',
       'AW.3', 'W', 'TEMP', 'DEWP', 'SLP', 'ALT', 'STP', 'MAX', 'MIN', 'PCP01',
       'PCP06', 'PCP24', 'PCPXX', 'SD'],
      dtype='object')


In [6]:
data.dtypes

USAF             int64
WBAN             int64
YR--MODAHRMN     int64
DIR             object
SPD             object
GUS             object
CLG             object
SKC             object
L               object
M               object
H               object
VSB             object
MW              object
MW.1            object
MW.2            object
MW.3            object
AW              object
AW.1            object
AW.2            object
AW.3            object
W               object
TEMP            object
DEWP            object
SLP             object
ALT             object
STP             object
MAX             object
MIN             object
PCP01           object
PCP06           object
PCP24           object
PCPXX           object
SD              object
dtype: object

## Problem 3 - Descriptive statistics
Based on the data DataFrame from Problem 1, answer to following questions:

What is the mean Fahrenheit temperature in the data (use the TEMP column)?

What is the standard deviation of the Maximum temperature (use the MAX column)?

How many unique stations exists in the data (use the USAF column)?

In [7]:
temp = data['TEMP']

In [8]:
temp= temp[pd.to_numeric(temp, errors='coerce').notnull()] #remove non-numeric rows
temp = temp.astype(float) #convert from str to floats
mean_F = print(temp.mean()) #get the mean

52.24796852279531


In [9]:
max_temp = data['MAX']

In [10]:
max_temp= max_temp[pd.to_numeric(max_temp, errors='coerce').notnull()] #remove non-numeric rows
max_temp = max_temp.astype(float) #convert from str to floats
std_F = print(max_temp.std()) #get the mean

10.30798909464196


In [11]:
usaf = data['USAF']

In [12]:
print(usaf.unique())

[28450 29980]


In [14]:
print(f'the number of unique stations is: {len(usaf.unique())}')

the number of unique stations is: 2


## Exercise 3.2
In this exercise, you will clean the data from our data file by removing no-data values, convert temperature values in Fahrenheit to Celsius, and split the data into separate datasets using the weather station identification code. We will start this problem by cleaning and converting our temperature data. An overview of the tasks in this exercise:

Create a new dataframe called selected that contains selected columns from the data file
Clean the new DataFrame by removing no-data values
Create a new column for temperatures converted from Fahrenheit to Celsius
Divide the data into separate DataFrames for the Helsinki Kumpula and Rovaniemi stations
Save the new DataFrames to CSV files

## Problem 1 - Read the data and remove NaN values
The first step for this problem is to read the data file 6153237444115dat.csv into a variable data using pandas and cleaning it a bit:

Select the columns USAF, YR--MODAHRMN, TEMP, MAX, MIN from the data DataFrame and assign them to a variable selected
Remove all rows from selected that have NoData in the column TEMP using the dropna() function

In [23]:
selected = data.loc[:, ['USAF', 'YR--MODAHRMN', 'TEMP', 'MAX', 'MIN']]

In [24]:
selected = selected.apply(pd.to_numeric, errors='coerce') #convert str to floats and if non-numeric to nan
selected = selected.dropna(subset=['TEMP']) #drop rows with nan in TEMP

## Problem 2 - Convert temperatures to Celsius
Convert the temperature values from Fahrenheits to Celsius:

Create a new column to selected called Celsius.
Convert the Fahrenheit temperatures from TEMP using the conversion formula below and store the results in the new Celsius column:
TempCelsius = (TempFahrenheit - 32) / 1.8
Round the values in the Celsius column to have 0 decimals (do not create a new column, update the current one)
Convert the Celsius values into integers (do not create a new column, update the current one)

In [25]:
selected ['Celsius'] = (selected['TEMP']-32)/1.8
selected.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN   Celsius
0  28450  201705010000  31.0  NaN  NaN -0.555556
1  28450  201705010020  30.0  NaN  NaN -1.111111
2  28450  201705010050  30.0  NaN  NaN -1.111111
3  28450  201705010100  31.0  NaN  NaN -0.555556
4  28450  201705010120  30.0  NaN  NaN -1.111111

In [26]:
selected ['Celsius'] = selected ['Celsius'].round(0)
selected.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0  28450  201705010000  31.0  NaN  NaN     -1.0
1  28450  201705010020  30.0  NaN  NaN     -1.0
2  28450  201705010050  30.0  NaN  NaN     -1.0
3  28450  201705010100  31.0  NaN  NaN     -1.0
4  28450  201705010120  30.0  NaN  NaN     -1.0

In [27]:
selected ['Celsius'] = selected ['Celsius'].astype('int')
selected.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0  28450  201705010000  31.0  NaN  NaN       -1
1  28450  201705010020  30.0  NaN  NaN       -1
2  28450  201705010050  30.0  NaN  NaN       -1
3  28450  201705010100  31.0  NaN  NaN       -1
4  28450  201705010120  30.0  NaN  NaN       -1

## Problem 3 - Select data and save to disk
Divide the data in selected into two separate DataFrames:

Select all rows from the selected DataFrame with the USAF code 29980 into a variable called kumpula.
Select all rows from the selected DataFrame with the USAF code 28450 into a variable called rovaniemi.
Save the kumpula DataFrame into a file Kumpula_temps_May_Aug_2017.csv in CSV format:
Separate the columns with commas (,)
Use only 2 decimals for the floating point numbers
Repeat the same procedures and save the rovaniemi DataFrame into a file Rovaniemi_temps_May_Aug_2017.csv.

In [21]:
selected['USAF'] = selected['USAF'].astype(str)
kumpula = selected[selected['USAF'].str.contains('29980')]
rovaniemi = selected[selected['USAF'].str.contains('28450')]

In [25]:
kumpula.to_csv('Kumpula_temps_May_Aug_2017.csv',decimal=',' )
rovaniemi.to_csv('Rovaniemi_temps_May_Aug_2017.csv',decimal=',' )

## Exercise 3.3
In this Exercise, we will explore our temperature data by comparing spring temperatures between Kumpula and Rovaniemi. To do this we'll use some conditions to extract subsets of our data and then analyse these subsets using basic pandas functions. Notice that in this exercise, we will use data saved from the previous Exercise (2.2.6), hence you should finish that Exercise before this one. An overview of the tasks in this exercise:

Calculate the median temperatures for Kumpula and Rovaniemi for the summer of 2017
Select temperatures for May and June 2017 in separate DataFrames for each location
Calculate descriptive statistics for each month (May, June) and location (Kumpula, Rovaniemi)

## Problem 1 - Read the data and calculate basic statistics
Read in the CSV files generated in Exercise 2.2.6 to the variables kumpula and rovaniemi and answer to following questions:

What was the median Celsius temperature during the observed period in Helsinki Kumpula? Store the answer in a variable kumpula_median.
What was the median Celsius temperature during the observed period in Rovaniemi? Store the answer in a variable rovaniemi_median.

In [28]:
kumpula = pd.read_csv('Kumpula_temps_May_Aug_2017.csv')
rovaniemi = pd.read_csv('rovaniemi_temps_May_Aug_2017.csv')
kumpula.shape

(2924, 7)

In [20]:
kumpula_median = kumpula['Celsius'].median()
kumpula_median

14.0

In [21]:
rovaniemi_median = rovaniemi['Celsius'].median()
rovaniemi_median

11.0

## Problem 2 - Select data and compare temperatures between months
The median temperatures above consider data from the entire summer (May-Aug), hence the differences might not be so clear. Let's now find out the mean temperatures from May and June 2017 in Kumpula and Rovaniemi:

From the kumpula and rovaniemi DataFrames, select the rows where values of the YR--MODAHRMN column are from May 2017. Assign these selected rows into the variables kumpula_may and rovaniemi_may
Repeat the procedure for the month of June and assign those values into variables to kumpula_june and rovaniemi_june
Calculate and print the mean, min and max Celsius temperatures for both places in May and June using the new subset dataframes (kumpula_may, rovaniemi_may, kumpula_june, and rovaniemi_june). Answer to following questions:
Does there seem to be a large difference in temperatures between the months?
Is Rovaniemi a much colder place than Kumpula?

In [31]:
kumpula

Unnamed: 0   USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0           8770  29980  201705010000  37,0  NaN  NaN        3
1           8771  29980  201705010100  37,0  NaN  NaN        3
2           8772  29980  201705010200  37,0  NaN  NaN        3
3           8773  29980  201705010300  37,0  NaN  NaN        3
4           8774  29980  201705010400  39,0  NaN  NaN        4
...          ...    ...           ...   ...  ...  ...      ...
2919       11689  29980  201708311900  64,0  NaN  NaN       18
2920       11690  29980  201708312000  64,0  NaN  NaN       18
2921       11691  29980  201708312100  64,0  NaN  NaN       18
2922       11692  29980  201708312200  64,0  NaN  NaN       18
2923       11693  29980  201708312300  64,0  NaN  NaN       18

[2924 rows x 7 columns]

In [36]:
kumpula_may =kumpula.loc[(kumpula['YR--MODAHRMN'] >= 201705010000) & (kumpula['YR--MODAHRMN'] < 201706010000)]
kumpula_may

Unnamed: 0   USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0          8770  29980  201705010000  37,0  NaN  NaN        3
1          8771  29980  201705010100  37,0  NaN  NaN        3
2          8772  29980  201705010200  37,0  NaN  NaN        3
3          8773  29980  201705010300  37,0  NaN  NaN        3
4          8774  29980  201705010400  39,0  NaN  NaN        4
..          ...    ...           ...   ...  ...  ...      ...
736        9506  29980  201705311900  51,0  NaN  NaN       11
737        9507  29980  201705312000  50,0  NaN  NaN       10
738        9508  29980  201705312100  47,0  NaN  NaN        8
739        9509  29980  201705312200  44,0  NaN  NaN        7
740        9510  29980  201705312300  43,0  NaN  NaN        6

[741 rows x 7 columns]

In [33]:
rovaniemi_may =rovaniemi.loc[(rovaniemi['YR--MODAHRMN'] >= 201705010000) & (rovaniemi['YR--MODAHRMN'] < 201706010000)]
rovaniemi_may

Unnamed: 0   USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0              0  28450  201705010000  31,0  NaN  NaN       -1
1              1  28450  201705010020  30,0  NaN  NaN       -1
2              2  28450  201705010050  30,0  NaN  NaN       -1
3              3  28450  201705010100  31,0  NaN  NaN       -1
4              4  28450  201705010120  30,0  NaN  NaN       -1
...          ...    ...           ...   ...  ...  ...      ...
2215        2217  28450  201705312220  32,0  NaN  NaN        0
2216        2218  28450  201705312250  32,0  NaN  NaN        0
2217        2219  28450  201705312300  33,0  NaN  NaN        1
2218        2220  28450  201705312320  32,0  NaN  NaN        0
2219        2221  28450  201705312350  30,0  NaN  NaN       -1

[2220 rows x 7 columns]

In [34]:
kumpula_june =kumpula.loc[(kumpula['YR--MODAHRMN'] >= 201706010000) & (kumpula['YR--MODAHRMN'] < 201707010000)]
rovaniemi_june =rovaniemi.loc[(rovaniemi['YR--MODAHRMN'] >= 201706010000) & (rovaniemi['YR--MODAHRMN'] < 201707010000)]

In [38]:
# this cell is is to get the mean max and min of Celcius 
kumpula_may.describe().iloc[[1,3,-1],3] 

mean     9.758435
min     -2.000000
max     23.000000
Name: Celsius, dtype: float64

In [39]:
kumpula_june.describe().iloc[[1,3,-1],3]

mean    13.732493
min      3.000000
max     24.000000
Name: Celsius, dtype: float64

In [41]:
rovaniemi_may.describe().iloc[[1,3,-1],3]

mean     2.986036
min     -7.000000
max     15.000000
Name: Celsius, dtype: float64

In [42]:
rovaniemi_june.describe().iloc[[1,3,-1],3]

mean    11.017395
min     -1.000000
max     23.000000
Name: Celsius, dtype: float64

In [43]:
#Does there seem to be a large difference in temperatures between the months?
# Is Rovaniemi a much colder place than Kumpula?
avg_temp_kumpula = (kumpula_may['Celsius'].mean() + kumpula_june['Celsius'].mean())/2
avg_temp_rovaniemi = (rovaniemi_may['Celsius'].mean() + rovaniemi_june['Celsius'].mean())/2
print(f' kumpula: {avg_temp_kumpula}, rovaniemi: {avg_temp_rovaniemi}')

 kumpula: 11.745463772553556, rovaniemi: 7.001715714303867


## yes there is a big difference. rovaneimi is colder


## Problem 3 - Parse daily temperatures by aggregating data
In this problem, the aim is to aggregate the hourly temperature data for Kumpula and Rovaniemi weather stations to a daily level. Currently, there are at most three measurements per hour in the data, as you can see from the YR--MODAHRMN column:

    USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0  28450  201705010000  31.0  NaN  NaN       -1
1  28450  201705010020  30.0  NaN  NaN       -1
2  28450  201705010050  30.0  NaN  NaN       -1
3  28450  201705010100  31.0  NaN  NaN       -1
4  28450  201705010120  30.0  NaN  NaN       -1
In this problem you should:

Summarize the information for each day by aggregating (grouping) the DataFrame using the groupby() function.
The output should be a new DataFrame where you have calculated mean, max and min Celsius temperatures for each day separately based on hourly values.
Repeat the task for the two data sets you created in Problem 2 (May-August temperatures from Rovaniemi and Kumpula).


In [44]:
gb = kumpula_may.groupby('YR--MODAHRMN')

In [39]:
kumpula_may

USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
8770  29980  201705010000  37.0  NaN  NaN        3
8771  29980  201705010100  37.0  NaN  NaN        3
8772  29980  201705010200  37.0  NaN  NaN        3
8773  29980  201705010300  37.0  NaN  NaN        3
8774  29980  201705010400  39.0  NaN  NaN        4
...     ...           ...   ...  ...  ...      ...
9506  29980  201705311900  51.0  NaN  NaN       11
9507  29980  201705312000  50.0  NaN  NaN       10
9508  29980  201705312100  47.0  NaN  NaN        8
9509  29980  201705312200  44.0  NaN  NaN        7
9510  29980  201705312300  43.0  NaN  NaN        6

[741 rows x 6 columns]

In [45]:
kumpula_may = kumpula_may.set_index('YR--MODAHRMN') #set index
kumpula_may.index = pd.to_datetime(kumpula_may.index, format='%Y%m%d%H%M') #index to datetime

In [46]:
# this cell splits the dataframe into a list with multiple dataframes based on day
df = kumpula_may.copy()
df = df.reset_index()
splits= list(df.groupby(df['YR--MODAHRMN'].dt.day))
splits

[(1,
            YR--MODAHRMN  Unnamed: 0   USAF  TEMP   MAX   MIN  Celsius
  0  2017-05-01 00:00:00        8770  29980  37,0   NaN   NaN        3
  1  2017-05-01 01:00:00        8771  29980  37,0   NaN   NaN        3
  2  2017-05-01 02:00:00        8772  29980  37,0   NaN   NaN        3
  3  2017-05-01 03:00:00        8773  29980  37,0   NaN   NaN        3
  4  2017-05-01 04:00:00        8774  29980  39,0   NaN   NaN        4
  5  2017-05-01 05:00:00        8775  29980  41,0   NaN   NaN        5
  6  2017-05-01 06:00:00        8776  29980  44,0  44,0  35,0        7
  7  2017-05-01 07:00:00        8777  29980  45,0   NaN   NaN        7
  8  2017-05-01 08:00:00        8778  29980  48,0   NaN   NaN        9
  9  2017-05-01 09:00:00        8779  29980  51,0   NaN   NaN       11
  10 2017-05-01 10:00:00        8780  29980  52,0   NaN   NaN       11
  11 2017-05-01 11:00:00        8781  29980  51,0   NaN   NaN       11
  12 2017-05-01 12:00:00        8782  29980  54,0   NaN   NaN       12
 

In [47]:
mean_value = []
max_value = []
min_value = []
kumpula_may_daily = []
for df in splits:
    mean_value = df[1]['Celsius'].mean() 
    max_value = df[1]['Celsius'].max() 
    min_value = df[1]['Celsius'].min() 
    row1 = [mean_value,max_value,min_value]
    kumpula_may_daily.append(row1)
    
kumpula_may_daily = pd.DataFrame(kumpula_may_daily, columns = ['mean', 'max', 'min'])
kumpula_may_daily

mean  max  min
0    7.625000   12    3
1    9.750000   16    2
2    9.208333   13    4
3    6.666667   11    3
4   10.250000   17    2
5   10.958333   15    6
6    5.875000    9    2
7    4.250000    7    1
8    1.666667    4   -2
9    1.541667    4   -2
10   2.708333    8   -2
11   5.166667   11   -2
12   7.583333   13    1
13  10.125000   14    3
14   8.708333   14    3
15   7.217391   12    1
16   9.708333   13    6
17  11.000000   14    9
18  16.375000   23    9
19  16.958333   22   13
20  14.833333   18   11
21  14.666667   19    9
22  13.416667   18    8
23  12.291667   17    9
24  14.041667   20    9
25  13.833333   18    8
26   9.416667   12    7
27  12.916667   18    8
28  13.636364   18    6
29   9.708333   13    4
30  10.625000   16    6

In [48]:
dates =[] #get date index from split
for df in splits:
    df = df[1].reset_index()
    date_at_each_day = df.loc[0,'YR--MODAHRMN']
    dates.append(date_at_each_day)

kumpula_may_daily.index = dates

In [49]:
kumpula_may_daily

mean  max  min
2017-05-01   7.625000   12    3
2017-05-02   9.750000   16    2
2017-05-03   9.208333   13    4
2017-05-04   6.666667   11    3
2017-05-05  10.250000   17    2
2017-05-06  10.958333   15    6
2017-05-07   5.875000    9    2
2017-05-08   4.250000    7    1
2017-05-09   1.666667    4   -2
2017-05-10   1.541667    4   -2
2017-05-11   2.708333    8   -2
2017-05-12   5.166667   11   -2
2017-05-13   7.583333   13    1
2017-05-14  10.125000   14    3
2017-05-15   8.708333   14    3
2017-05-16   7.217391   12    1
2017-05-17   9.708333   13    6
2017-05-18  11.000000   14    9
2017-05-19  16.375000   23    9
2017-05-20  16.958333   22   13
2017-05-21  14.833333   18   11
2017-05-22  14.666667   19    9
2017-05-23  13.416667   18    8
2017-05-24  12.291667   17    9
2017-05-25  14.041667   20    9
2017-05-26  13.833333   18    8
2017-05-27   9.416667   12    7
2017-05-28  12.916667   18    8
2017-05-29  13.636364   18    6
2017-05-30   9.708333   13    4
2017-05-31  10.625000   16    6